java/arcs/core/util/UnionFind.kt

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.core.util
+
+/**
+ * A union-find datastructure for computing equivalence classes
+ * consisting of `Element` instances.
+ */
+class UnionFind<Element> {
+    /**
+     * A node in the union find datastructure.
+     */
+    private inner class Node(e: Element, p: Node?) {
+        val element: Element = e
+        var parent: Node? = p
+    }
+
+    private val nodes = HashMap<Element, Node>()
```

Comment:
prefer the more idiomatic `mutableMapOf<Element, Node>()`, which will use a HashMap behind the scenes